packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,432 @@
+    SELECT startPos, endPos, size, hash
+    FROM transcriptSpans
+    WHERE vatID = ? AND current = 1
+  `);
+
+  /**
+   * Obtain the bounds and other metadata for a vat's current transcript span.
+   *
+   * @param {string} vatID  The vat in question
+   *
+   * @returns {{startPos: number, endPos: number, size: number, hash: string}}
+   */
+  function getCurrentSpanBounds(vatID) {
+    const bounds = sqlGetCurrentSpanBounds.get(vatID);
+    assert(bounds, `no current transcript for ${vatID}`);
+    return bounds;
+  }
+
+  function spanArtifactName(rec) {
+    return `transcript.${rec.vatID}.${rec.startPos}.${rec.endPos}`;
+  }
+
+  function historicSpanMetadataKey(rec) {
+    if (rec.current) {
+      return `export.transcript.${rec.vatID}.current`;
+    } else {
+      return `export.transcript.${rec.vatID}.${rec.startPos}`;
+    }
+  }
+
+  function currentSpanMetadataKey(rec) {
+    return `export.transcript.${rec.vatID}.current`;
+  }
+
+  function spanRec(vatID, startPos, endPos, hash, size) {
+    return { vatID, startPos, endPos, hash, size };
+  }
+
+  const sqlEndCurrentSpan = db.prepare(`
+    UPDATE transcriptSpans
+    SET current = 0
+    WHERE current = 1 AND vatID = ?
+  `);
+
+  /**
+   * End the current transcript span for a vat and start a new one.
+   *
+   * @param {string} vatID  The vat whose transcript is to rollover to a new
+   *    span.
+   */
+  function rolloverSpan(vatID) {
+    ensureTxn();
+    const { hash, startPos, endPos, size } = getCurrentSpanBounds(vatID);
+    const rec = spanRec(vatID, startPos, endPos, hash, size);
+    noteExport(historicSpanMetadataKey(rec), JSON.stringify(rec));
+    sqlEndCurrentSpan.run(vatID);
+    sqlWriteSpan.run(vatID, endPos, endPos, initialHash, 1, 0);
+  }
+
+  const sqlGetAllSpanMetadata = db.prepare(`
+    SELECT vatID, startPos, endPos, hash, current, size
+    FROM transcriptSpans
+    ORDER BY vatID, startPos
+  `);
+
+  const sqlGetCurrentSpanMetadata = db.prepare(`
+    SELECT vatID, startPos, endPos, hash, current, size
+    FROM transcriptSpans
+    WHERE current = 1
+    ORDER BY vatID, startPos
+  `);
+
+  /**
+   * Obtain artifact metadata records for spans contained in this store.
+   *
+   * @param {boolean} includeHistorical If true, include all spans that are
+   * present in the store regardless of their currency; if false, only include
+   * the current span for each vat.
+   *
+   * @yields {[key: string, value: string]}
+   * @returns {Iterable<[key: string, value: string]>}  An iterator over pairs of
+   *    [historicSpanMetadataKey, rec], where `rec` is a JSON-encoded metadata record for the
+   *    span named by `historicSpanMetadataKey`.
+   */
+  function* getExportRecords(includeHistorical) {
+    const sql = includeHistorical
+      ? sqlGetAllSpanMetadata
+      : sqlGetCurrentSpanMetadata;
+    for (const rec of sql.iterate()) {
+      yield [historicSpanMetadataKey(rec), JSON.stringify(rec)];
```

Comment:
Yes. Sqlite does the right thing.  When you specify the columns in the `SELECT` statement you get just the columns requested (you might get more with `SELECT *` but we're not doing that) and we use `ORDER BY` to make sure the ordering of the records is invariant.

Reply:
I don't understand what `ORDER BY` has anything to do with this. The concern is the number and order of the fields in the record, not the order of export records relative to each other, which doesn't matter in this case.

In particular `spanRec` excludes `current` but both `sqlGetAllSpanMetadata` and `sqlGetCurrentSpanMetadata` include the `current` field in theit `SELECT`. Furthermore, if the order in which fields appear in the `SELECT` statement changes, it risk being different than the one in which they appear in `spanRec`. This is a major refactoring hazard.

Reply:
I did a little experimentation, and determined that the order that the properties get added to the result record (which is the same as the order they are output by `JSON.stringify`) does not depend on the order the column names appear in the `SELECT` statement but on the order the columns are declared in the original `CREATE TABLE` statement (presumably because the authors of the `better-sqlite3` package figured they were returning a JS object and so the order didn't matter since it was name based rather than positional).

That does seem like a refactoring hazard, so I added an intermediate take-it-apart-put-it-back-together operation to ensure controlled ordering.

An open question is whether or not the metadata should include the `isCurrent` property.  I assumed that it should, and so included it on purpose, but, like `size`, it's not actually needed for correct operation so we can take it out if that's the consensus.

Reply:
It turns out I was mistaken in one point there -- `isCurrent` is needed for import to work properly.  So it stays in.